출저 : 주니온TV(유튜브)

## 콜라츠의 추측: 우박수 문제

 - 쉬운데 어렵다. 대학에서 잘 가르킨다.
 ![image-20210927120303710](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210927120303710.png)

 ![image-20210927120351739](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210927120351739.png)


 

### 구현(collatz) : 콜라츠 수열을 만들어 반환
 - 증명은 못할정도로 어렵지만, 구현은 쉽다.
 ![image-20210927120854123](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210927120854123.png)

In [4]:
def collatz(n):
    # 0. n부터 나누기2로 줄거나 or *3+1로 커지는데, 그것을  n==1될때까지 모은다.
    seq = [n]
    # 1. 콜라츠는 1이되면 종료한다. 1보다 클때까지만 돌린다.
    while n>1:
        if n%2==0:
            n=n//2
        else:
            n=3*n + 1
        seq.append(n)
    return seq


collatz(10)

[10, 5, 16, 8, 4, 2, 1]

In [ ]:
for i in range(1, 10):
    print(collatz(i))

### 규칙발견을 위해, seq-len을 end=" "로 출력해보자.
 - 수열의 길이에 규칙이 있는지 봤는데, 안보인다...

In [ ]:
for i in range(1, 100):
    print(len(collatz(i)), end=" ")

### 규칙은 없지만, 쉬워서 문제는 우박수처럼 쏟아진다.

![image-20210927121646260](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210927121646260.png)

- 손으로는 계산못한다.

![image-20210927121732546](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210927121732546.png)



 - 조건을 만족시키는 가장 작은 수는? 
     - **input index가 정해지지 않은 경우, 그냥  while True:로 무한루프  i=0, i+=1로 시작하자.**
     - 1부터 시작하며, +1씩 나가면서, 조건을 만족할 때 바로 return(break)해버리면 `가장 작은 수`다.
        - **두번째 수는? count변수를 만들어 조건 만족할때마다 +1 해서 체크한다.**
     - 끝을 모르는 경우 while True: + break(or return)의 무한루프를 돈다.

In [9]:
# 콜라츠 수열의 길이가 100인 숫자들 중에서 가장 작은 수는?
n=0
while True:
    n+=1
    if len(collatz(n)) == 100:
        break 

print("길이가 100인 가장 작은 우박수는?", n)

길이가 100인 가장 작은 우박수는? 322


In [16]:
# 10000이하의 콜라츠 수열의 길이가 가장긴 수는?
# -> 1부터 10000까지를 돌아야한다 -> for문으로가능 / 너무 많지 않기 때문에 다돌아도 됨.
# -> 돌면서 여러개 중에 가장 긴것 찾기 -> for문위에 변수에 curr변수 update
# --> update 당시의 i도 알아야한다면...  curr = max(curr, ??? )로 한줄업데이트 불가능.ㅠ -> if문으로 풀어서 + 변수도 챙겨야한다.

max_len = 0
max_halestone = []

for n in range(1, 10000+1):
     #curr_len = max(curr_len, len(collatz(i)))
     # 만약, 수열자체도 필요하다면, 
     halestone = collatz(n)
     length = len(halestone)
     if max_len < len(halestone):
         max_len = length
         max_halestone = halestone


print("10,000이하 가장 긴 우박수의 길이와 우박수 :", max_len, max_halestone[0])

10,000이하 가장 긴 우박수의 길이와 우박수 : 262 6171


### 연습문제: 콜라츠수열의 길이가 500보다 큰 첫번째 자연수는?
 - 콜라츠수열의 길이가 500보다 큰 첫번째 자연수는?
     - 첫번째 자연수 -> 조건을 만족하면 바로 return한 수
 - 콜라츠수열의 길이가 500보다 큰 10번째 자연수의  콜라츠 수열 길이는?
     - 10번재 자연수 -> 조건을 만족할 때마다 count따로 해줘야함.

In [22]:
n=0
count = 0
while True:
    n+=1
    if len(collatz(n)) > 500:
        count+=1
        print(count, n)
        if count ==10:
            break

n

1 626331
2 704623


KeyboardInterrupt: 

## 로마 문자(기호) -> 아라비아 숫자로 

![image-20210928165007845](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210928165007845.png)

![image-20210928165014112](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210928165014112.png)



### 로마 숫자 표기법

![image-20210928165213769](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210928165213769.png)

![image-20210928165245431](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210928165245431.png)

- X, XX, XXX : 10, 20, 30
  - **XL : 40**



![image-20210928165353548](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210928165353548.png)

- L, LX, LXX, LXXX : 50, 60, 70, 80
  -  **XC : 90**



![image-20210928165541773](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210928165541773.png)

- C, CC, CCC : 100 200 300 
  - **CD : 400**



![image-20210928165645098](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210928165645098.png)

- D, DC, DCC, DCCC: 500, 600, 700, 800
  - **MC : 900**

![image-20210928165730181](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210928165730181.png)

- 규칙발견하기
  - **기본적으로는 `가산표기법` : 기호의 값을 그냥 합산하면 됨. **
  - **너무 길어지는 경우, `감산표기법`**: 
    - **큰 수 앞에(왼쪽에) 작은 수를 넣으면, 큰 수(오른)에서  작은 수(왼)를 빼라는 뜻.**
    - 예) 4, 9, 40, 90, 400, 900 ..





In [25]:
string = "MDCLXVI"

print(string[0])
print(string[6])

for i in range(len(string)):
    print(string[i], end=" ")
print()

print("M" + "D" + "C" + "L" + "X" + "V" + "I")

M
I
M D C L X V I 
MDCLXVI


### getValueOf(symbol) : 로마 문자를 아라비아 숫자로 변환

In [31]:
# 함수로 매칭..
def getValueOf(symbol):
    if symbol == 'M': return 1000
    elif symbol == 'D': return 500
    elif symbol == 'C': return 100
    elif symbol == 'L': return 50
    elif symbol == 'X': return 10
    elif symbol == 'V': return 5
    elif symbol == 'I': return 1

roman = "MDCLXVI"

for i in range(len(roman)):
    #print(getValueOf(roman[i]))
    print(roman[i] ,":", getValueOf(roman[i]))



M : 1000
D : 500
C : 100
L : 50
X : 10
V : 5
I : 1


### dict로 매칭되는 데이터 매핑
 - 함수로 길게 if elif...로 매칭을 시켜 return해줄 필요가 없다.

In [35]:
romanDict = {
    'M':1000, 'D':500, 'C':100, 'L':50, 'X':10, 'V':5, 'I':1
}

roman = "MDCLXVI"

for i in range(len(roman)):
    #print(getValueOf(roman[i]))
    print(roman[i] ,":", romanDict[roman[i]])

M : 1000
D : 500
C : 100
L : 50
X : 10
V : 5
I : 1


### 단순 가산만 해주기
 - 가산표기법은 고려안한 상태로 로마문자->숫자변환->단순 가산만
 - **어차피 숫자로 변환후 각 대응되는 숫자를 누적해서 더하던지 빼던지 해야함 -> 변수에 누적합**

In [37]:
roman = "MDCLXVI"

arabic = 0

for i in range(len(roman)):
    #arabic += getValueOf(roman[i])
    arabic += romanDict[roman[i]]

print(arabic)


1666


### 감산 생각해주기
 - i번째 수에서, i+1(다음수)보다 작은 숫자라고 판명되면 arabic에 -= 빼주면 된다.
     - for문에 i+1이 들어온 순간,, 바깥  for문에서는 1개 적게 끝나고 + 맨 마지막꺼는 추가로 더해준다?
     - 또는, 여기서는, 감산 확인 if문 자체를... 마지막 인덱스는 포함안되도록, i <( n-1)로 추가 조건주기


In [41]:
def toArabicNumber(roman):
    arabic = 0

    for i in range(len(roman)):
        # 마지막 [n-1]에서는 검사안하도록.. and 조건을 주자. 
        # - 그외에는.. 다 검사하게 되어있다..
        if (i< len(roman)-1) and romanDict[roman[i]] < romanDict[roman[i+1]]:
            arabic -= romanDict[roman[i]]
        else:
            arabic += romanDict[roman[i]]

    return arabic


romans = ['CCLXIX', 'LXXX', 'XXIX']
arabics =[]
for roman in romans:
    arabics.append(toArabicNumber(roman))

print(arabics)

[269, 80, 29]


### 연습문제: 아라비아 숫자를 로마문자로 변환가능할까?
 - 참고 : www.wolframalpha.com
     - `roman numerals 1666`
     - `roman numerals MDCLXVI`

## 아라비아숫자를 로마문자(숫자)로

 - my) 단위숫자로 제거하는데, 단위와 단위사이에 감산법이 걸릴 때 먼저 감산법으로 제거해준다.

![image-20210930162015769](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210930162015769.png)

In [42]:
romanDict

{'M': 1000, 'D': 500, 'C': 100, 'L': 50, 'X': 10, 'V': 5, 'I': 1}

In [43]:
# 이번에는 거꾸로 매핑
romanDict = {
    1:'I',
    5:'V',
    10:'X',
    50:'L',
    100:'C',
    500:'D',
    1000:'M',
}

In [44]:
print(romanDict.keys())
print(romanDict.values())
print(romanDict.items())

dict_keys([1, 5, 10, 50, 100, 500, 1000])
dict_values(['I', 'V', 'X', 'L', 'C', 'D', 'M'])
dict_items([(1, 'I'), (5, 'V'), (10, 'X'), (50, 'L'), (100, 'C'), (500, 'D'), (1000, 'M')])


In [45]:
for key in romanDict.keys():
    print(key,":", romanDict[key])

1 : I
5 : V
10 : X
50 : L
100 : C
500 : D
1000 : M


### 1부터 9까지만 아라비아->로마문자로 변환해보기
 - n을 받으면, 1까지 돌아가면서, romanDict에서 꺼내서 처리해보기
 - 중간중간 예외가 생긴다.

In [49]:
# 작성하고 보니까..
# 1) 감산법 적용 숫자들은 따로 먼저 찾아서 처리해주고 그만큼 감산
# 2) 5 등의 단위숫자들은 들어있는 만큼 감산 (N>=단위 로 확인해서 가진것만큼 제한다.)
# 3) 단위숫자들 다끝나고 나머지 숫자들(5이하의 수) I의 1로 갯수만큼 돌린다.
def romanNumber1(n):

    str = ""

    while n>=1 : # n<1 즉, n=0일때 종료된다. while 조건 -> not 조건에서 종료 // 조건 -> while not조건 동안 계속
        # 감산법이 적용되는 놈들.. 4, 9, 40, 90, 등을 먼저 처리해줘야한다.
        if n==9:
            # 9면, IX를 출력해야한다. -> 1로 가져와야하고, 10으로 가져와야한다.
            str += romanDict[1] + romanDict[10]
            # 가져왔으면 그만큼.. n에서 빼줘야한다..
            n-=9 
        elif n==4: # IV
            str += romanDict[1] + romanDict[5]
            n-=4
        elif n>=5: # 4, 9의 감산이 아닌 상태에서의 5이상.. 5,6,7,8  V + (n-5) * I 가 아니라..
            # 5를 로마자로 V로 변환하여 추가해주고, 5만 감산시키는 작전.. 나머지는 나중에 처리하도록 n에 남겨둔다.
            # 단위로서, 반복적으로 제거되도록한다?
            str += romanDict[5] 
            n-=5
        else:
            # 5보다 작은 것들은 그 갯수만큼.. I를 반복해서 더해주면 된다.
            for _ in range(n):
                str += romanDict[1]
            n-=n 


    return str

romanNumber1(8)

'VIII'

In [50]:
for n in range(1, 10):
    print(n, ":", romanNumber1(n))

1 : I
2 : II
3 : III
4 : IV
5 : V
6 : VI
7 : VII
8 : VIII
9 : IX


### 100미만의 숫자로 확장하기
 - 1~9까지는 romanNumber1()로 처리하는 것도 포함된다.

In [52]:
# 1. 단위숫자 가지고 있는 것부터 거름
# 1-1. 단위숫자 거르기 전에 그 범위내 감산법 대상부터처 처리, [n>=50] 4, 9, 40, [90]
# 1-1. 단위숫자 제거 
# 2. 그 다음 단위숫자 가지고 거름
# 2-1. 감산법..
# 자리수가 바뀌면 그 때의 함수로 처리해서 return받는다.
def romanNumber2(n):
    str = "" 
    # 100미만에서 단위숫자인 50을 들고 있으면, 50을 안가질때까지 제해야하므로 while로 가져온다.
    while n>=50:
        # 50이상을 가지고 있어도, 40, 90 등 감산부터 처리한다.
        if n>=90: # 90 -> 10 + 100 로 먼저 처리한다.
            str += romanDict[10] + romanDict[100]
            n -= 90
        else:
            # 88이나 가지고 있다면.. 감산법 없으면, ㅇ단위 숫자부터 제한다.
            str += romanDict[50]
            n-=50 
    # 감산, 및 단위숫자 제거가 되서 50미만이 됨.

    # 그 다음 단위숫자를 처리한다.
    while n>=10 : 
        if n>=40:
            str += romanDict[10] + romanDict[50]
            n-=40
        else:
            str += romanDict[10]
            n-=10
    
    # 10~99까지 처리되었으면, 1~9의 함수결과와 더한다.
    return str + romanNumber1(n)

romanNumber2(99)


'XCIX'

In [53]:
# 10, 20 ~ 90까지 해보자.
for n in range(10, 100, 10):
    print(n, ":", romanNumber2(n))

10 : X
20 : XX
30 : XXX
40 : XL
50 : L
60 : LX
70 : LXX
80 : LXXX
90 : XC


### 4000미만의 숫자로 더 확장하기
 - 단위가 5000이상은 없다고 치기 때문에 3999로 제한함.. 4000이면 5000 - 1000으로 감산되어야함.

In [55]:
def romanNumber3(n):
    
    str = ""

    while n>=1000:
        # 이놈은 상위에 존재하는 감산법(4000)이 없기 때문에.. 단위숫자만 제한다.
        str += romanDict[1000]
        n-=1000

    while n>=500:
        # 900감산해주기
        if n>=900:
            str += romanDict[100] + romanDict[1000]
            n-=900
        else:
            str += romanDict[500]
            n-=500

    while n>= 100:
        if n>=400:
            str += romanDict[100] + romanDict[500]
            n-=400
        else:
            str += romanDict[100]
            n -= 100 
    return str + romanNumber2(n)

for n in range(100, 1000, 100):
    print(n, ":", romanNumber3(n))

100 : C
200 : CC
300 : CCC
400 : CD
500 : D
600 : DC
700 : DCC
800 : DCCC
900 : CM


### 패턴이 보인다.

 - while마다 패턴이 있으니, 그 패턴에 맞게 필요한 요소들을 list로 구성해서 만들어보자.

![image-20210930170716150](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210930170716150.png)

- null string에다가 범위(단위를 가지고 있는) 동안에는

  - 감산법

  - 단위 반복 제거

    

    ![image-20210930170813702](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210930170813702.png)
    ![image-20210930170846688](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210930170846688.png)



- dict에서 꺼내쓰는 대신 (**key:value를    index가 동일한 위치에 각각을 list1:list2로 선언해서 사용**)
  - **각 패턴에 들어가는 숫자들이 짝을 이루도록** list의 **같은 index에 나열**해서 사용한다.
    - whlie에 들어갈 단위    [1000,      500,   100, ...]
    - if에 들어갈 감산법처리 [4000(X), 900,   400, ...]
    - str+=에 들어갈 감산법용 로마숫자 900이면 100(C)과 1000(M)가 같은 index line에 오도록 list 2개를 나열
      ![image-20210930171524960](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210930171524960.png)
      - 이부분을 지정하는 것이 까다로우니까.. 걍 같은 index에 나열해서 임의의 쌍을 지어놓는 것임.
        ![image-20210930171617831](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210930171617831.png)
        ![image-20210930171637933](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210930171637933.png)
        ![image-20210930171825096](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210930171825096.png)









![image-20210930170945904](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210930170945904.png)

### dict로 모든 것을 다 지정해놓고 사용하면 더 간단해진다.
 - 단위뿐만 아니라 감산법의 범위도 list에 다 몰아놓고, 그 때의 로마문자도 dict에 다 몰아놓은 상태에서 while 패턴을 돌린다.
 ![image-20210930172105516](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210930172105516.png)

 - 테스트
  ![image-20210930172320015](https://raw.githubusercontent.com/is3js/screenshots/main/image-20210930172320015.png)




In [57]:
romanDict2 = {
    1:'I', 4:'IV',
    5:'V', 9:'IX',
    10:'X', 40:'XL',
    50:'L', 90:'XC',
    100:'C', 400:'CD',
    500:'D', 900:'CM',
    1000:'M',
}

keyValues = [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1]

In [60]:
# 감산법구간도.. 단위구간처럼  while >=  그만큼 제하기를 할 수 있다.
# 1000과 500사이 900이 등장하면.. 이미 1000을 다 돌린 상태기 때문에.. 999이하라서 900을 while에 돌리는 것이 가능하다.
# - 비록 단위숫자가 아니면 1번밖에 해당이 안되지만.. while에 들어가도 상관없다 1번만 수행하고 나오면 되므로.

def toRomanNumber(n):
    str=""
    for i in range(len(keyValues)):
        while n >= keyValues[i]:
            str += romanDict2[keyValues[i]]
            n -= keyValues[i]

    return str 


toRomanNumber(155)

'CLV'